File: templates/matlab/chord_diagram.m
function fig = chord_diagram()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(7);
    n = 6;
    F = double(randi([0 8], n, n));
    F(F < 3) = 0;
    F(1:n+1:end) = 0;
    totals = sum(F, 1)' + sum(F, 2);
    gap = 0.05;
    span = (2*pi - n*gap)*totals/sum(totals);
    starts = pi/2 + [0; cumsum(span(1:end-1) + gap)];
    seg_o = zeros(n, n, 2); seg_i = zeros(n, n, 2);
    cur = starts;
    for i = 1:n
        for j = 1:n
            if F(i, j) > 0
                w = F(i, j)/totals(i)*span(i);
                seg_o(i, j, :) = [cur(i), cur(i)+w]; cur(i) = cur(i) + w;
            end
        end
        for j = 1:n
            if F(j, i) > 0
                w = F(j, i)/totals(i)*span(i);
                seg_i(i, j, :) = [cur(i), cur(i)+w]; cur(i) = cur(i) + w;
            end
        end
    end
    fig = figure('Position', [100 100 560 540]); hold on;
    for i = 1:n                                % ribbons
        for j = 1:n
            if F(i, j) > 0
                a1 = seg_o(i, j, 1); a2 = seg_o(i, j, 2);
                b1 = seg_i(j, i, 1); b2 = seg_i(j, i, 2);
                poly = [arc_pts(a1, a2, 0.96); chord_pts(a2, b1, 0.96); ...
                        arc_pts(b1, b2, 0.96); chord_pts(b2, a1, 0.96)];
                fill(poly(:, 1), poly(:, 2), palette('cat', i), ...
                     'FaceAlpha', 0.45, 'EdgeColor', 'none');
            end
        end
    end
    for k = 1:n                                % node arc bands + labels
        band = [arc_pts(starts(k), starts(k)+span(k), 1.0); ...
                arc_pts(starts(k)+span(k), starts(k), 1.09)];
        fill(band(:, 1), band(:, 2), palette('cat', k), 'EdgeColor', 'none');
        mid = starts(k) + span(k)/2;
        text(1.26*cos(mid), 1.26*sin(mid), sprintf('N%d', k), ...
             'HorizontalAlignment', 'center', 'FontSize', 9);
    end
    xlim([-1.55 1.55]); ylim([-1.55 1.55]);
    axis equal; axis off;
    title('Chord diagram of flows', 'FontSize', 11);
end

function p = arc_pts(t1, t2, r)
    a = linspace(t1, t2, 40)';
    p = [r*cos(a), r*sin(a)];
end

function p = chord_pts(t1, t2, r)
    p0 = r*[cos(t1), sin(t1)];
    p1 = r*[cos(t2), sin(t2)];
    t = linspace(0, 1, 40)';
    p = (1-t).^2.*p0 + t.^2.*p1;
end

File: templates/python/chord_diagram.py
"""chord_diagram: 和弦图（6 节点流量矩阵，弧带宽=流量，贝塞尔连接）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle


def _arc(t1, t2, r, m=40):
    a = np.linspace(t1, t2, m)
    return np.c_[r * np.cos(a), r * np.sin(a)]


def _chord(t1, t2, r, m=40):
    p0 = r * np.array([np.cos(t1), np.sin(t1)])
    p1 = r * np.array([np.cos(t2), np.sin(t2)])
    t = np.linspace(0, 1, m)[:, None]
    return (1 - t)**2 * p0 + t**2 * p1


def make_figure(F=None, title='Chord diagram of flows'):
    apply_theme()
    rng = np.random.default_rng(7)
    n = 6
    if F is None:
        F = rng.integers(0, 9, (n, n)).astype(float)
        F[F < 3] = 0
        np.fill_diagonal(F, 0)
    totals = F.sum(0) + F.sum(1)
    gap = 0.05
    span = (2 * np.pi - n * gap) * totals / totals.sum()
    starts = np.pi / 2 + np.concatenate(([0], np.cumsum(span[:-1] + gap)))
    seg_o = np.zeros((n, n, 2)); seg_i = np.zeros((n, n, 2))
    cur = starts.copy()
    for i in range(n):
        for j in range(n):
            if F[i, j] > 0:
                w = F[i, j] / totals[i] * span[i]
                seg_o[i, j] = (cur[i], cur[i] + w); cur[i] += w
        for j in range(n):
            if F[j, i] > 0:
                w = F[j, i] / totals[i] * span[i]
                seg_i[i, j] = (cur[i], cur[i] + w); cur[i] += w
    fig, ax = plt.subplots(figsize=(5.6, 5.4))
    ax.set_xlim(-1.55, 1.55); ax.set_ylim(-1.55, 1.55)
    ax.set_aspect('equal'); ax.axis('off')
    for i in range(n):                      # ribbons
        for j in range(n):
            if F[i, j] > 0:
                a1, a2 = seg_o[i, j]; b1, b2 = seg_i[j, i]
                poly = np.vstack([_arc(a1, a2, 0.96), _chord(a2, b1, 0.96),
                                  _arc(b1, b2, 0.96), _chord(b2, a1, 0.96)])
                ax.fill(poly[:, 0], poly[:, 1], color=cycle(i),
                        alpha=0.45, lw=0, zorder=1)
    for k in range(n):                      # node arc bands + labels
        band = np.vstack([_arc(starts[k], starts[k] + span[k], 1.0),
                          _arc(starts[k] + span[k], starts[k], 1.09)])
        ax.fill(band[:, 0], band[:, 1], color=cycle(k), lw=0, zorder=3)
        mid = starts[k] + span[k] / 2
        ax.text(1.26 * np.cos(mid), 1.26 * np.sin(mid), f'N{k+1}',
                ha='center', va='center', fontsize=9)
    ax.set_title(title, fontsize=11)
    fig.tight_layout()
    return fig


if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)
